### ISSUE TITLE
NeighborLoader not sampling edge labels correctly

### ISSUE THREAD

**Issue by nabihach** [pyg-team/pytorch_geometric#4351]
### 🐛 Describe the bug

I'm trying to use GCNs to do link prediction on the OGB_MAG dataset as follows. (The script is minimal working example, but is long because I had to add reverse edges and then create train/val/test split for edges): 

```
import torch
import os
from torch_geometric.datasets import OGB_MAG
from torch_geometric.data import Data
from torch_geometric.loader import NeighborLoader
from torch_geometric.nn import GCNConv
import torch_geometric.transforms as T

path = os.path.join('..', 'data', 'OGB_MAG')
mag_dataset = OGB_MAG(path, preprocess='metapath2vec')
mag_data = mag_dataset[0]
```
```
# Add reverse edges to the heterogeneous graph.

r, c = mag_data.edge_index_dict[('author', 'affiliated_with', 'institution')]
mag_data[('institution', 'to', 'author')].edge_index = torch.stack([c, r])
r, c = mag_data.edge_index_dict[('author', 'writes', 'paper')]
mag_data[('paper', 'to', 'author')].edge_index = torch.stack([c, r])
r, c = mag_data.edge_index_dict[('paper', 'has_topic', 'field_of_study')]
mag_data[('field_of_study', 'to', 'paper')].edge_index = torch.stack([c, r])
r, c = mag_data.edge_index_dict[('paper', 'cites', 'paper')]
mag_data[('paper', 'revcites', 'paper')].edge_index = torch.stack([c, r])
```
```
# Split data into train/val/set

tfs = T.RandomLinkSplit(is_undirected=False, 
                      add_negative_train_samples=True,
                      disjoint_train_ratio=0.0,# disjoint mode if > 0
                      split_labels=False,
                      edge_types=[('author', 'affiliated_with', 'institution'), ('author', 'writes', 'paper'), ('paper', 'cites', 'paper'), ('paper', 'has_topic', 'field_of_study')], 
                      rev_edge_types=[('institution', 'to', 'author'), ('paper', 'to', 'author'), ('paper', 'revcites', 'paper'), ('field_of_study', 'to', 'paper')], 
                      )
train_data, val_data, test_data = tfs(mag_data)
```
```
# Create training data for link prediction task

train_x= torch.cat(  (train_data['paper'].x, train_data['author'].x, train_data['field_of_study'].x, train_data['institution'].x), 0).float()
train_y = torch.cat( (train_data[('paper', 'cites', 'paper')].edge_label, train_data[('author', 'writes', 'paper')].edge_label, train_data[('paper', 'has_topic', 'field_of_study')].edge_label, train_data[('author', 'affiliated_with', 'institution')].edge_label),  0)
train_edge_index = torch.cat( (train_data[('paper', 'cites', 'paper')].edge_label_index, train_data[('author', 'writes', 'paper')].edge_label_index, train_data[('paper', 'has_topic', 'field_of_study')].edge_label_index, train_data[('author', 'affiliated_with', 'institution')].edge_label_index),  1)

training_data = Data(x=train_x, y=train_y, edge_index=train_edge_index)
training_data = T.ToSparseTensor(remove_edge_index=False)(training_data)

print(training_data)
>>> Data(x=[1939743, 128], edge_index=[2, 29555416], y=[29555416], adj_t=[1939743, 1939743, nnz=29555416])
```
```
# Create model

hidden_channels = 128
model = GCNConv(hidden_channels, hidden_channels, cached=False)
```
```
# Create neighbor loader

input_nodes = torch.ones(train_data.num_nodes, dtype=torch.bool)
kwargs = {'batch_size': 1024, 'num_workers': 6, 'persistent_workers': True}
train_loader = NeighborLoader(training_data, num_neighbors=[10] * 2, shuffle=True, directed=True, input_nodes=input_nodes, **kwargs)
```
```
batch = next(iter(train_loader))
print(batch)
>>> Data(x=[35912, 128], edge_index=[2, 35790], y=[29555416], adj_t=[35912, 35912, nnz=35790], batch_size=1024)
```

As we can see, `batch.x` and `batch.edge_index` are the same size as `training_data.x` and `training_data.edge_index`. However, `batch.y` is the same size as `training_data.y`, which seems wrong. Any insights?

### Environment

* PyG version: 2.0.5
* PyTorch version: 1.10.0
* OS: Linux
* Python version: 3.8
* CUDA/cuDNN version: 10.2
* How you installed PyTorch and PyG: PyTorch from pip, PyG from source on 3/26/2022
* Any other relevant information: `torch.scatter==2.0.9`, `torch-sparse==0.6.13`, `torch-cluster==1.6.0`

**Comment by rusty1s** [comment]
Thanks for reporting. This is now fixed, see https://github.com/pyg-team/pytorch_geometric/commit/dc1569618b390a223deb148aac4243a5a0322166.
